src/jit/morph.cpp

```diff
@@ -12503,31 +12503,44 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
 
             /* See if we can fold GT_ADD nodes. */
 
-            if (oper == GT_ADD)
+            if (oper == GT_ADD || oper == GT_OR)
             {
-                /* Fold "((x+icon1)+(y+icon2)) to ((x+y)+(icon1+icon2))" */
+                /* Fold "((x+icon1)+(y+icon2)) to ((x+y)+(icon1+icon2))" same for bitwise OR */
 
-                if (op1->gtOper == GT_ADD && op2->gtOper == GT_ADD && !gtIsActiveCSE_Candidate(op2) &&
-                    op1->AsOp()->gtOp2->gtOper == GT_CNS_INT && op2->AsOp()->gtOp2->gtOper == GT_CNS_INT &&
-                    !op1->gtOverflow() && !op2->gtOverflow())
+                if (op1->OperIs(op2->OperGet()) && op1->OperIs(oper) && !gtIsActiveCSE_Candidate(op2) &&
+                    op1->AsOp()->gtGetOp2()->IsCnsIntOrI() && op2->AsOp()->gtGetOp2()->IsCnsIntOrI())
                 {
                     // Don't create a byref pointer that may point outside of the ref object.
                     // If a GC happens, the byref won't get updated. This can happen if one
                     // of the int components is negative. It also requires the address generation
                     // be in a fully-interruptible code region.
-                    if (!varTypeIsGC(op1->AsOp()->gtOp1->TypeGet()) && !varTypeIsGC(op2->AsOp()->gtOp1->TypeGet()))
+                    if (!varTypeIsGC(op1->AsOp()->gtGetOp1()->TypeGet()) &&
+                        !varTypeIsGC(op2->AsOp()->gtGetOp1()->TypeGet()))
                     {
-                        cns1 = op1->AsOp()->gtOp2;
-                        cns2 = op2->AsOp()->gtOp2;
-                        cns1->AsIntCon()->gtIconVal += cns2->AsIntCon()->gtIconVal;
+                        cns1          = op1->AsOp()->gtGetOp2();
+                        cns2          = op2->AsOp()->gtGetOp2();
+                        ssize_t icon1 = cns1->AsIntCon()->IconValue();
+                        ssize_t icon2 = cns2->AsIntCon()->IconValue();
+
+                        if (oper == GT_ADD)
+                        {
+                            if (op1->gtOverflow() || op2->gtOverflow())
+                            {
+                                break;
+                            }
+                            cns1->AsIntCon()->SetIconValue(icon1 + icon2);
+                        }
+                        else
+                        {
+                            cns1->AsIntCon()->SetIconValue(icon1 | icon2);
```

Comment:
Should have an assert that the oper is correct:  GT_OR

Reply:
That follows from the surrounding if-clause